packages/headless/src/app/engine.ts

```diff
@@ -0,0 +1,248 @@
+import {Logger} from 'pino';
+
+type EngineDispatch<State> = ThunkDispatch<
+  State,
+  ThunkExtraArguments,
+  AnyAction
+> &
+  Dispatch<AnyAction>;
+
+export interface Engine<State = SearchAppState> {
+  /**
+   * Dispatches an action directly. This is the only way to trigger a state change.
+   * Each headless controller dispatches its own actions.
+   *
+   * @param action An action supported by the headless engine.
+   *
+   * @returns For convenience, the action object that was just dispatched.
+   */
+  dispatch: EngineDispatch<State>;
+  /**
+   * Adds a change listener. It will be called any time an action is
+   * dispatched, and some part of the state tree may potentially have changed.
+   * You may then access the new `state`.
+   *
+   * @param listener A callback to be invoked on every dispatch.
+   * @returns A function to remove this change listener.
+   */
+  subscribe: (listener: () => void) => Unsubscribe;
+  /**
+   * The complete headless state tree.
+   */
+  state: State;
+  /**
+   * The redux store.
+   */
+  store: Store;
+  /**
+   * A function for headless to call to retrieve a refreshed access token.
+   */
+  renewAccessToken: () => Promise<string>;
+  /**
+   * The logger instance used by headless.
+   * */
+  logger: Logger;
+  /**
+   * Adds the specified reducers to the store.
+   */
+  addReducers(reducers: ReducersMapObject): void;
+}
+
+export interface EngineOptions<Reducers extends ReducersMapObject> {
+  /**
+   * The global headless engine configuration options.
+   */
+  configuration: EngineConfigurationOptions;
+  /**
+   * Map object of reducers.
+   * A reducer is a pure function that takes the previous state and an action, and returns the next state.
+   * ```
+   * (previousState, action) => nextState
+   * ```
+   * [Redux documentation on reducers.](https://redux.js.org/glossary#reducer)
+   */
+  reducers: Reducers;
+  /**
+   * The initial headless state.
+   * You may optionally specify it to hydrate the state
+   * from the server in universal apps, or to restore a previously serialized
+   * user session.
+   */
+  preloadedState?: StateFromReducersMapObject<Reducers>;
+  /**
+   * List of additional middlewares.
+   * A middleware is a higher-order function that composes a dispatch function to return a new dispatch function.
+   * It is useful for logging actions, performing side effects like routing, or turning an asynchronous API call into a series of synchronous actions.
+   * ```
+   * type MiddlewareAPI = { dispatch: Dispatch, getState: () => State }
+   * type Middleware = (api: MiddlewareAPI) => (next: Dispatch) => Dispatch
+   * ```
+   * [Redux documentation on middlewares.](https://redux.js.org/glossary#middleware)
+   */
+  middlewares?: Middleware<{}, StateFromReducersMapObject<Reducers>>[];
+  loggerOptions?: LoggerOptions;
+}
+
+interface SearchAPIClientArgument {
+  searchAPIClient: SearchAPIClient;
+}
+
+export function buildEngine<Reducers extends ReducersMapObject>(
```

Comment:
Just thinking out loud here, but for this function `buildEngine`, is there a meaningful UT that we could add ? 

Something that verifies that the core functionality are respected ?

That it update basic config, that analytics config is performed, stuff like that.

Reply:
Yep absolutely, we have these tests inside `headless-engine.test.ts`. I will move them to a new `engine.test.ts` file.